xonsh/environ.py

```diff
@@ -1135,9 +1137,36 @@ def load_static_config(ctx, config=None):
     return conf
 
 
+def _splitpath(path, sofar=[]):
+    folder, path = os.path.split(path)
+    if path == "":
+        return sofar[::-1]
+    elif folder == "":
+        return (sofar + [path])[::-1]
+    else:
+        return _splitpath(folder, sofar + [path])
+
+_CHARACTER_MAP = {chr(o): '_%s' % chr(o+32) for o in range(65, 91)}
+_CHARACTER_MAP.update({'.': '_.', '_': '__'})
+
+
+def _cache_renamer(path):
+    path = os.path.abspath(path)
+    o = [''.join(_CHARACTER_MAP.get(i, i) for i in w) for w in _splitpath(path)]
+    o[-1] = "{}.{}".format(o[-1], sys.implementation.cache_tag)
+    return o
+
+
+def _make_if_not_exists(dirname):
+    if not os.path.isdir(dirname):
+        os.makedirs(dirname)
+
+
 def xonshrc_context(rcfiles=None, execer=None):
     """Attempts to read in xonshrc file, and return the contents."""
     loaded = builtins.__xonsh_env__['LOADED_RC_FILES'] = []
+    homedir = builtins.__xonsh_env__['HOME']
+    cachedir = os.path.join(homedir, '.xonshrc_cache')
```

Comment:
I think that `$XONSH_DATA_DIR/xonshrc_cache` (see [here](http://xon.sh/envvars.html#xonsh-data-dir)) is probably a better location that in the home directory.